### ISSUE TITLE
Large Memory Leaks on iOS-JS due to Spidermonkey

### ISSUE THREAD

**Issue by DavidDeSimone** [cocos2d/cocos2d-x#16327]
- cocos2d-x version: 3.11
- devices test on: iPad Air 2
- developing environments
  - NDK version: N/A
  - Xcode version: 7.2.1
  - VS version: N/A
  - brower type and version:

We have discovered several large and concerning memory leaks originating from spidermonkey in our Cocos2d-x project on the iOS platform. We have not confirmed if this issue is or isn't happening on other platforms due to lack of good memory leak tooling on our other development platform (Android). Our rough polling of Heap Size in my Android application seems to suggest that this issue is also happening on Android, but we don't have solid proof of this. The two principle leaks seem to originate from the following two actions:
- Calling eval (either directly or indirectly) on any script code will leak the string of the evaluated code.
- Calling a function that has undergone cc.Class's _super variable replacement will cause the "ArgumentsData" object of Spidermonkeys Internal "ArgumentsObject" to leak. ArgumentsObject here is spidermonkeys internal C++ representation of the "arguments" object in a Javascript function. 

We believe this is related to issue https://github.com/cocos2d/cocos2d-x/issues/16291, and that issue is caused by these leaks. 

Examining either of these action's in Apple's Instruments program produces the following data:

These are the stacks for the Eval case
![screen shot 2016-08-05 at 2 17 41 pm](https://cloud.githubusercontent.com/assets/7245174/17450856/9dbd03c4-5b17-11e6-8e26-5edec404f2d7.png)
![screen shot 2016-08-05 at 2 17 19 pm](https://cloud.githubusercontent.com/assets/7245174/17450868/a15ae46a-5b17-11e6-9cc6-54caf81a8df3.png)

This is the stack for the arguments case
![screen shot 2016-08-05 at 2 17 19 pm](https://cloud.githubusercontent.com/assets/7245174/17450888/c5303bce-5b17-11e6-80d6-b0126c42f0cc.png)

For the eval() case
Steps to Reproduce:
1. Examine a Cocos2d-x application with the Instruments program, examining memory leaks.
2. Repeated eval a string of arbitrary Javascript.
3. Watch as leaked memory rises.

For the _super case
Steps to Reproduce:
1. Examine a Cocos2d-x application with the Instruments program, examining memory leaks.
2. Run the following code:

```
var HelloWorld = cc.Layer.extend({
                    ctor:function () {
                        this._super();
                        this.scheduleUpdate();
                    },
                    update: function () {
                        var node = new cc.Node();
                        this.addChild(node);
                        this.scheduleOnce(function () {
                            node.removeFromParent();
                        });
                    }
                });

```
1. Watch as leaked memory rises.
2. Now, if you were to create an update function that does not undergo super replacement, this leak will not occur.

**Comment by ricardoquesada** [comment]
Thanks for the info!
Just to make sure, are you calling the garbage collector after calling `eval` and the `_super` thing?

**Comment by DavidDeSimone** [comment]
So we have an update:

It turns out that "leaks" is unreliable for profiling spidermonkey. Source: https://developer.mozilla.org/en-US/docs/Mozilla/Performance

If you look at the passage for "Leaks":
`Apple provides some tools for Mac OS X that report similar problems to those reported by LSAN and Valgrind. The "leaks" tool is not recommended for use with SpiderMonkey or Firefox, because it gets confused by tagged pointers and thinks objects have leaked when they have not (see bug 390944).`. The bug confirms that "Leaks" cannot be used to profile spidermonkey, and reports a large number of false positives. 

However, in the process, we were able to find issues with the way that we are restarting the scripting core via Director::getInstance()->restart(). We were mistaking this leak for what "leaks" was reporting to us above. Calling Director::getInstance()->restart() will leak the javascript global object. It will "leak" this in the sense that the global object will not be garbage collected. We can confirm this by observing that the finalize callback for the global object in scripting core (sc_finalize) does not fire when you call __restartVM() in the javascript. We found that the cause of this is how we are managing the persisted rooted objects in our js_type_class_t structs. We are wrapping the roots in a mozilla::Maybe object, which is not being properly cleaned up in our restart flow. Calling the maybe's destroyIfConstructed() method does not seem to be properly cleaning up the root either, for reasons we don't fully understand. If we bypass the maybe object, and just manage the persisted root via calls to new and delete, the global object does indeed get cleaned up. 

So for Repro, set up a JS project, call __restartVM, notice that the log in sc_finalize is not being hit.

**Comment by pandamicro** [comment]
@DavidDeSimone Thanks for the insight of restart issue, we were using Maybe because PersistentRootObject can't be declared in the header directly and can't be created with the constructor of ScriptingCore, but as you said, now I think we can declare it as `JS::PersistentRootedObject *` and use new/delete to manage it. I will check this out.

I have also found the reason of #16291, it's actually not a leak, it's triggered by a memory release delay in JSB.

**Comment by DavidDeSimone** [comment]
Hello @pandamicro 

Thank you for looking into this. I am going to submit a PR with our work around. You may choose not to use it, or fix this issue in another way, but I figure it can save you some effort in diving into the potential problem.

**Comment by DavidDeSimone** [comment]
I have submitted the PR that supplies our code here. https://github.com/cocos2d/cocos2d-x/pull/16383

**Comment by pandamicro** [comment]
@DavidDeSimone Thanks for your help, our implementation was actually the same, but I missed the `js_type_class_t`, so just merged your commit into my PR https://github.com/cocos2d/cocos2d-x/pull/16380.

**Comment by DavidDeSimone** [comment]
Closing this issue as it was fixed in 3.13
